lib/uiauto-client.js

```diff
@@ -0,0 +1,237 @@
+// The Command Proxy relays UIAuto message to and from Appium. It is also the
+// UIAuto facade for Appium.
+//
+// The message route is the following:
+// Appium <--> Command Proxy <--> Instruments
+// The medium between Instruments and Command Proxy is the command-proxy-client
+// script.
+//
+// Command Proxy --> Instruments message format: {cmd:"<CMD>"}
+//
+// Instruments --> Command Proxy message format:
+// <one char message type>,<stringified json data>
+// <stringified json data> format:
+// {status:<status>, value:<result>}
+
+import log from './logger';
+import _fs from 'fs';
+import through from 'through';
+import net from 'net';
+import _mkdirp from 'mkdirp';
+import path from 'path';
+import Promise from 'bluebird';
+import _ from 'lodash';
+
+let fs = {
+  unlink: Promise.promisify(_fs.unlink)
+};
+let mkdirp = Promise.promisify(_mkdirp);
+
+
+var MORE_COMMAND = '#more';
+var MESSAGE_TYPES = ['error', 'no data', 'regular', 'chunk', 'last chunk'];
+
+var UNKNOWN_ERROR = {
+  status: 13,
+  value: 'Error parsing socket data from instruments'
+};
+
+
+class UIAutoClient {
+  constructor (opts = {}) {
+    this.curCommand = null;
+    this.onReceiveCommand = null;
+    this.commandQueue = [];
+    this.sock = opts.sock || '/tmp/instruments_sock';
+    this.socketServer = null;
+    this.hasConnected = false;
+    this.bufferedData = '';
+    this.currentSocket = null;
+    this.firstSocketConnectionCb = null;
+    this.resultBuffer = '';
+  }
+
+  isChunk (data) {
+    return data.type && data.type.indexOf('chunk') >= 0;
+  }
+
+  getResultAndSendNext (data, c) {
```

Comment:
also, i'm not sure how, but i feel that there must be a cleaner way to implement this that doesn't rely on so much state and also makes use of async/await perhaps.